Common/Securities/Option/QLOptionPriceModel.cs

```diff
@@ -0,0 +1,244 @@
+                    }
+                    catch (Exception err)
+                    {
+                        Log.Debug("tryGetImpliedVol() error: " + err.Message);
+                        return 0m;
+                    }
+                };
+
+                Func<Tuple<decimal, decimal>> evalDeltaGamma = () =>
+                {
+                    try
+                    {
+                        return Tuple.Create((decimal)option.delta(), (decimal)option.gamma());
+                    }
+                    catch (Exception)
+                    {
+                        if (optionSecurity.EnableGreekApproximation)
+                        {
+                            var step = 0.01;
+                            var initial = underlyingQuoteValue.value();
+                            underlyingQuoteValue.setValue(initial - step);
+                            var npvMinus = EvaluateOption(option);
+                            underlyingQuoteValue.setValue(initial + step);
+                            var npvPlus = EvaluateOption(option);
+                            underlyingQuoteValue.setValue(initial);
+
+                            return Tuple.Create((decimal)((npvPlus - npvMinus) / (2 * step)),
+                                                (decimal)((npvPlus - 2 * npv + npvMinus) / (step * step)));
+                        }
+                        else
+                            return Tuple.Create(0.0m, 0.0m);
+                    }
+                };
+
+                Func<double> reevalVega = () =>
+                {
+                    var step = 0.001;
+                    var initial = underlyingVolValue.value();
+                    underlyingVolValue.setValue(initial + step);
+                    var npvPlus = EvaluateOption(option);
+                    underlyingVolValue.setValue(initial);
+
+                    return (npvPlus - npv) / step;
+                };
+
+                Func<double> reevalTheta = () =>
+                {
+                    var step = 1.0 / 365.0;
+
+                    Settings.setEvaluationDate(settlementDate.AddDays(-1));
+                    var npvMinus = EvaluateOption(option);
+                    Settings.setEvaluationDate(settlementDate);
+
+                    return (npv - npvMinus) / step;
+                };
+
+                Func<double> reevalRho = () =>
+                {
+                    var step = 0.001;
+                    var initial = riskFreeRateValue.value();
+                    riskFreeRateValue.setValue(initial + step);
+                    var npvPlus = EvaluateOption(option);
+                    riskFreeRateValue.setValue(initial);
+
+                    return (npvPlus - npv) / step;
+                };
+
+                // producing output with lazy calculations of IV and greeks
+
+                return new OptionPriceModelResult((decimal)npv,
```

Comment:
I'm testing this branch at home and even after setting `EnableGreekApproximation = true`, they are all always 0 because this line throws an exception: `NPV()` is always returning `NaN` for some reason causing this line to throw.  Would something else be possible here or to catch the NaN case in `EvaluateOption`?